File: java/src/main/java/com/java/heap/Heap.java
package com.java.heap;

public class Heap<E extends Comparable> implements PQInterface<E> {
	private E[] A;
	private int numItems;

	public Heap(int arraySize) { // 기본 초기 생성자
		A = (E[]) new Comparable[arraySize]; // 여기는 Comparable로 선언하고..
		numItems = 0;
	}
	
	
	public Heap(E[] B, int numElements) { // 바깥에서 배열을 받은 경우의 생성자
		A = B;
		numItems = numElements;
	}

	@Override
	public void insert(E newItem) throws PQException {
		if (numItems < A.length) {
			A[numItems] = newItem;
			percolateUp(numItems);
			numItems++;
		} else throw new PQException("HeapErr: Insert()-Overflow!");
	}

	// 스며오르기
	public void percolateUp(int i) {
		int parent = (i-1)/2;
		if (parent >= 0 && A[i].compareTo(A[parent])>0) { //비교
			E tmp = A[i];
			A[i] = A[parent];
			A[parent] = tmp;
			percolateUp(parent); // 계속 이어가는 경우
		}
	}


	@Override
	public E deleteMax() throws PQException {
		if (!isEmpty()) {
			E maxElement = A[0];
			A[0] = A[numItems - 1];
			numItems--;
			percolateDown(0);
			return maxElement;
		} else throw new PQException("HeapErr: DeleteMax()-Underflow");

	}


	public void percolateDown(int i) {
		int leftChild = (2 * i) + 1;
		int rightChild = (2 * i) + 2;
		int child = leftChild; // 디폴트 차일드

		if (child <= numItems - 1) { // 탈출
			if (rightChild <= numItems - 1 && A[leftChild].compareTo(A[rightChild]) < 0) {
				child = rightChild;
			}
			if (A[i].compareTo(A[child]) < 0) {
				E tmp = A[i];
				A[i] = A[child];
				A[child] = tmp;
				percolateDown(child);
			}

		}
	}


	public void buildHeap() {
		if (numItems >= 2) {
			for (int i = (numItems - 2) / 2; i >= 0; i--) {
				percolateDown(i); // i가 parent값이 되도록 해야하는데 -1하면 계산이 안맞누
			}
		}
	}


	@Override
	public E max() throws PQException {
			if (!isEmpty()) {
				return A[0];
			} else throw new PQException("HeapErr: Max()-empty");
		}


	@Override
	public boolean isEmpty() {
		return numItems == 0;
	}

	@Override
	public void clear() {
		A = (E[]) new Object[A.length]; // 여기는 Object로 선언하고..
		numItems = 0;
	}
}

File: kotlin/src/main/kotlin/com/kotlin/heap/Heap.kt
package com.kotlin.heap


class Heap(sizeOfArray: Int = 6) : PQInterface {
    private var arr: IntArray
    private var numItems: Int
    private var sizeOfArray = 6

    // 초기 생성자
    init {
        arr = IntArray(sizeOfArray)
        numItems = 0
        this.sizeOfArray = sizeOfArray
    }

    // 보조 생성자 : 바깥에서 배열을 받은 경우
    constructor(arr: IntArray, numItems:Int): this() {
        this.arr = arr
        this.numItems = numItems
        this.sizeOfArray = numItems
    }

    // 삽입
    override fun insert(newItem: Int) {
        if (numItems < sizeOfArray) {
            arr[numItems] = newItem
            percolateUp(numItems)
            numItems++
        } else throw PQException("HeapErr: Insert()-Overflow!")

    }

    // 후 스며오르기
    private fun percolateUp(i: Int) {
        var tmp = 0
        val parent = (i - 1) / 2

        if ((parent >= 0) && (arr[i] > arr[parent])) {
            tmp = arr[i]
            arr[i] = arr[parent]
            arr[parent] = tmp
            percolateUp(parent)
        }
    }


    override fun deleteMax(): Int {
        if (!isEmpty()) {
            val maxElement = arr[0]
            arr[0] = arr[numItems-1]
            numItems--
            percolateDown(0)
            return maxElement
        } else throw PQException("HeapErr: DeleteMax()-Underflow")
    }


    fun percolateDown(i: Int) {
        var tmp = 0
        var child = (2 * i )+ 1
        val rightChild = (2 * i) + 2

        if (child <= numItems - 1) {
            if (rightChild <= numItems -1 && arr[child] < arr[rightChild]) {
                child = rightChild
            }
            if (arr[i] < arr[child]) {
                tmp = arr[i]
                arr[i] = arr[child]
                arr[child] = tmp
                percolateDown(child)
            }
        }
    }


    override fun max(): Int {
        if (!isEmpty()) return arr[0]
        else throw PQException("HeapErr: Max()-empty")
    }


    // 무조건 루트부터 시작, i=0으로 시작하게끔
    override fun buildHeap() {
        if (numItems >= 2) { // 노드가 2개 이상부터 되야
            for (i in 0 until (numItems - 2) / 2 step 1) { // 이거 잘보기
                percolateDown(i)
            }
        }
    }


    override fun isEmpty(): Boolean {
        return (numItems == 0)
    }


    override fun clear() {
        arr = IntArray(numItems)
        numItems = 0
    }


}